### PR TITLE
#825: Allow passing explicit connection to 'Bucket.{list_blobs,make public}'

### PR DESCRIPTION
**Pull request by tseaver**
Requires #853.

See #825.

### CODE REVIEW

**Comment by dhermes** [comment]
Ping me on this when we resolve #853

**Comment by tseaver** [comment]
Rebased against updated branch for #853.

**Comment by tseaver** [comment]
Rebased again against updated branch for #853.

**Comment by tseaver** [comment]
@dhermes any remaining issues?

**Comment by dhermes** [comment]
Ditto here, was waiting on diffbase.

**Comment by tseaver** [comment]
@dhermes I rebased it this morning on the latest from #853.

**Comment by tseaver** [comment]
OIC, you'd like a rebase against master.

File: gcloud/storage/bucket.py [hunk 1]
```diff
@@ -60,12 +60,21 @@ class _BlobIterator(Iterator):
...
+    :param extra_params: Extra query string parameters for the API call.
+
+    :type connection: :class:`gcloud.storage.connection.Connection`
+    :param connection: The connection to use when sending requests.  Defaults
+                       to the bucket's connection
     """
-    def __init__(self, bucket, extra_params=None):
+    def __init__(self, bucket, extra_params=None, connection=None):
+        if connection is None:
+            connection = bucket.connection
```
Comment (dhermes) [inline gcloud/storage/bucket.py]: `_require_connection`

File: gcloud/storage/bucket.py [hunk 2]
```diff
@@ -854,18 +869,26 @@ def make_public(self, recursive=False, future=False):
...
-            doa.save()
+            doa.save(connection=connection)
 
         if recursive:
-            for blob in self:
+            for blob in self.list_blobs(projection='full',
+                                        connection=connection):
                 blob.acl.all().grant_read()
-                blob.save_acl()
+                blob.acl.save(connection=connection)
```
Comment (dhermes) [inline gcloud/storage/bucket.py]: This points out just how much this method is doing.

`for blob in self.list_blobs` is just insane for us to be looping through. It's certainly something we should put a cap on so people don't try to loop through millions of blobs. (Outside the scope of this PR, but worth pointing out.)

Comment (tseaver) [inline gcloud/storage/bucket.py]: I don't think `make_public` is insane for its intended usecase, which is to build out a static website and then publish it all at once.  Such a site isn't going to have millions of blobs.

**Comment by tseaver** [comment]
Ideally, the backend would have a special-purpose ACL for blobs where they "inherit"/"acquire" the permissions from their bucket;  one could then leave them in that state as they were created, and toggle them all public / private just by manipulating the bucket's ACL.

File: gcloud/storage/bucket.py (hunk 2 above)
Comment (dhermes) [inline gcloud/storage/bucket.py]: (Still somewhat outside the scope of this PR, but I'll keep going)

It's somewhat like `Bucket.delete(force=True)`. It works when it works, but can be a really bad idea if a user doesn't understand what they are doing.

- If we see this just as a method for buckets used as websites, we should rename
- We should strongly consider capping the number of blobs that can be iterated through, and maybe allow it to be overridden as an argument to the method

**Comment by dhermes** [comment]
Unfortunately that ideal doesn't exist.

File: gcloud/storage/bucket.py (hunk 2 above)
Comment (tseaver) [inline gcloud/storage/bucket.py]: Are these concerns blocking for this PR?

Comment (dhermes) [inline gcloud/storage/bucket.py]: Nope

**Comment by dhermes** [comment]
LGTM

**Pull request merged**